analysis/src/python/data_analysis/templates/filter_using_diff.py

```diff
@@ -0,0 +1,134 @@
+import argparse
+import sys
+from typing import List, Tuple
+
+import pandas as pd
+from diff_match_patch import diff_match_patch
+
+from analysis.src.python.data_analysis.model.column_name import StepColumns, SubmissionColumns
+from analysis.src.python.data_analysis.templates.utils.template_utils import parse_template_code_from_step
+from analysis.src.python.data_analysis.utils.code_utils import split_code_to_lines
+from analysis.src.python.data_analysis.utils.report_utils import parse_report
+from analysis.src.python.evaluation.tools.model.report import BaseIssue, BaseReport
+from analysis.src.python.utils.df_utils import filter_df_by_iterable_value, read_df, write_df
+from analysis.src.python.utils.logging_utils import configure_logger
+
+
+def get_issues_with_offset(report: BaseReport, code_lines: List[str]) -> List[Tuple[int, BaseIssue]]:
+    """ For each issue calculate offset in code written in one line and return as pairs sorted in ascending order."""
+
+    code_prefix_length = [0]
+    for code_line in code_lines:
+        code_prefix_length.append(code_prefix_length[-1] + len(code_line))
+
+    issues_with_offset = []
+    for issue in report.get_issues():
+        issue_offset = code_prefix_length[issue.get_line_number() - 1] + issue.get_column_number()
+        issues_with_offset.append((issue_offset, issue))
+
+    issues_with_offset.sort(key=lambda i: i[0])
+
+    return issues_with_offset
+
+
+def get_template_to_code_diffs(template_lines: List[str], code_lines: List[str]) -> List[Tuple[int, int, int]]:
+    """
+    Get template to students code diff in format of list of tuples (tag, start, end) where:
+    tag - type of change where 0 (code not changed), 1 (code added). Tag -1 (code deleted) is ignored.
+    start - position of change start (included)
+    end - position of change end (excluded)
+    """
+
+    matcher = diff_match_patch()
+    patches = matcher.diff_main(''.join(template_lines), ''.join(code_lines))
+
+    diffs = []
+    start = 0
+
+    for tag, patch in patches:
+        if tag != -1:
```

Comment:
I think we need to create an enum class with tags and use everywhere

Reply:
Yes, sure) Thanks!